iothub_client/tests/iothubclientcore_ut/iothubclientcore_ut.c

```diff
@@ -3587,6 +3587,59 @@ TEST_FUNCTION(IoTHubClientCore_UploadMultipleBlocksToBlobAsync_callbackEx_invoke
 
     IoTHubClientCore_Destroy(iothub_handle);
 }
+
+TEST_FUNCTION(IoTHubClientCore_UploadMultipleBlocksToBlobAsync_callbackEx_invoked_on_error_return_after_success)
+{
+    ///arrange
+    IOTHUB_CLIENT_CORE_HANDLE iothub_handle = IoTHubClientCore_Create(TEST_CLIENT_CONFIG);
+    TEST_GETDATA_CALLBACK_ARGS context = { 0 };
+    umock_c_reset_all_calls();
+
+    set_expected_calls_for_allocateUploadToBlob();
+    STRICT_EXPECTED_CALL(ThreadAPI_Create(IGNORED_PTR_ARG, IGNORED_PTR_ARG, IGNORED_PTR_ARG));
+    STRICT_EXPECTED_CALL(Lock(IGNORED_PTR_ARG));
+
+    STRICT_EXPECTED_CALL(singlylinkedlist_add(IGNORED_PTR_ARG, IGNORED_PTR_ARG)); /*this is adding HTTPWORKER_THREAD_INFO to the list of HTTPWORKER_THREAD_INFO's to be cleaned*/
+    STRICT_EXPECTED_CALL(ThreadAPI_Create(IGNORED_PTR_ARG, IGNORED_PTR_ARG, IGNORED_PTR_ARG));
+    STRICT_EXPECTED_CALL(Unlock(IGNORED_PTR_ARG));
+
+    STRICT_EXPECTED_CALL(get_time(IGNORED_NUM_ARG)).CallCannotFail();
+    STRICT_EXPECTED_CALL(IoTHubClientCore_LL_UploadMultipleBlocksToBlobEx(IGNORED_PTR_ARG, IGNORED_PTR_ARG, IGNORED_PTR_ARG, IGNORED_PTR_ARG))
+        .SetReturn(IOTHUB_CLIENT_OK);
+
+    STRICT_EXPECTED_CALL(IoTHubClientCore_LL_UploadToBlob_NotifyCompletion(IGNORED_PTR_ARG, IGNORED_PTR_ARG, IGNORED_PTR_ARG, IGNORED_PTR_ARG))
```

Comment:
it looks like `IoTHubClientCore_LL_UploadToBlob_NotifyCompletion` is not currently a mockable function, and this prevents the test from compiling. I think I can modify its declaration, but before putting more work into that I will wait for an initial review to ensure that the general approach of this patch is correct, and maybe if you have suggestions on a better way to handle this (maybe one function called by this one is already mocked, for example?)

Reply:
Well, it turns out I had a typo here, this should have been:

```
STRICT_EXPECTED_CALL(IoTHubClient_LL_UploadToBlob_NotifyCompletion(IGNORED_PTR_ARG, 
IGNORED_PTR_ARG, IGNORED_NUM_ARG, IGNORED_NUM_ARG, IGNORED_PTR_ARG))
         .SetReturn(IOTHUB_CLIENT_ERROR);
```

This fixes the build, but the test still doesn't work because the entire IoTHubClientCore_LL_UploadMultipleBlocksToBlobEx funcion is stubbed. I have modified the test to try to trace the functions used by IoTHubClientCore_LL_UploadMultipleBlocksToBlobEx; but I didn't manage to get that working either. Again I appreciate your help and guidance on how to make this test work.